game-core/src/test/java/games/strategy/triplea/delegate/battle/MockBattleState.java

```diff
@@ -0,0 +1,28 @@
+package games.strategy.triplea.delegate.battle;
+
+import games.strategy.engine.data.Unit;
+import java.util.Collection;
+import java.util.List;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+
+/**
+ * Simple implementation of BattleState for tests to use
+ *
+ * <p>{@link #givenBattleState()} will return a builder with everything defaulted and the test can
+ * override the specific items needed.
+ */
+@Builder
+public class MockBattleState implements BattleState {
```

Comment:
Nit, this is not really a Mock. A stub is something that returns the same value (the stub controls the return value), a mock is something that returns the value you tell it to (test controls the return value), and a 'fake' is a simplified implementation (aka, a test double).

In this case it looks like we basically have a real implementation. It's tempting to call this a test-fake, but given the API is so simple it looks to be full-blown. 

This makes me wonder if we should have a `DefaultBattleState` on the `BattleState` interface itself.  Essentially in `BattleState` declaring a public static class `DefaultBattleState` that would be this class. You've mentioned though you expect this class to grow and perhaps then we'll want to use a true-mock.

In the meantime the naming is a bit misleading and I think would be useful to fix. Perhaps we could call this a `DefaultBattleState`, or a `SimpleBattleState`.

Reply:
I'm trying to build a faked `BattleState` in a fluent manner where the test methods can easily modify the object to get it to the state they need.

With this mechanism, I could have one test that does
```
givenBattleState().attackingUnits(units).defendingUnits(units2).build();
```
while another test could do
```
givenBattleState().bombardingUnits(units).build();
```
and another could do
```
givenBattleState().bombardingUnits(units).battleSite(territory).build();
```

That's why I'm returning the `MockBattleStateBuilder` class instead of the `BattleStateBuilder` class.  I could add `@Builder(toBuilder = true)` and then do
```
givenBattleState().toBuilder().bombardingUnits(units).battleSite(territory).build();
```
but that felt like extra work that I could skip.

I'd like to use `mock()` but I don't see how I can do that in a fluent manner.  Instead, I'd have to do
```
BattleState battleState = givenBattleState();
when(battleState.attackingUnits()).thenReturn(units);
when(battleState.defendingUnits()).thenReturn(units2);
```

I could call it a fake because it will not be a full blown implementation.  Once I start adding in the updater methods, I'll not be implementing them.  The tests will just spy on the BattleState and check that the methods were called correctly.  And if I ever come across a step that updates the state and then reads from the updated state, I'll have rethink this :)  I haven't yet seen one but I haven't fully investigated every single step yet.

Reply:
I've renamed the class to `FakeBattleState` and the given method to `givenBattleStateBuilder()`.

Reply:
Understood on the fluent API, that is cool :+1: 

One word of caution/note, spying on state changes is generally a test code smell. It's generally better if one module computes the state change and returns the new state value. Then have another module that uses the first one to do the state mutation.

IE:
```

BiFunction<GameData, BattleState, Integer> computationStrategy;
BattleState battleState;

void foo(GameData gameData, BattleState battleState) {
  int newValue = computationStrategy.apply(gameData, battleState);
  battleState.setSomeValue(newValue);
}
```

In this case the test for the above is quite trivial as the computation function can be injected. In TripleA so far the cleanest way I know to do that is to use a static build function, ie:
```
@Builder
class ImperativeShell {
  @Nonnull private final BattleState battleState;
  @Nonnull private final GameData gameData;
  @Nonnull private final BiFunction<BattleState, GAmeDAta, Integer> computationFunction;

   public static ImperativeShell build(GameData gameData, BattleState battleState) {
         return builder()
                      .battleState(battleState)
                      .gameData(gameData)
                      .computationFunction(new RealComputationFunction())
                      .build();
   }
}
```

Then in test you can use the builder() directly to inject mocks.

At the end of this you can then test the 'RealComputationFunction' directly and assert it returns the expected values without having to spy for side-effects. It turns assertions that look like this:

```
computation.apply(battleState);

verify(battleState).unitRemoved(unit);
```

Into:
```
Unit result = computation.apply(battleState);

assertThat(result, is(someExpectedUnit));
```

Reply:
Nota bene, if the objects to inject are all unique minus the generic types (recall that in java generic types are dropped during compilation time), then `@InjectMocks` can make a test even easier and automatically link mocks to an `@RequiredsArgsConstructor` (which is created automatically when using `@Builder`), ie:
```
class ImperativeShellTest {
  @Mock private BattleState battleState;
  @Mock private GameData gameData;
  @Mock private BiFunction<BattleState, GAmeDAta, Integer> computationFunction;
   
   @InjectMocks
   private ImperativeShell imperativeShell;
```